### PR TITLE
Kafkamdm sasl mechanism

### PR DESCRIPTION
**Pull request by Antoine-Auffret**
Add for kafkamdm route the sasl mechanism to match the recent support of kafka sasl in metrictank https://github.com/grafana/metrictank/pull/1956

### CODE REVIEW

**Comment by Antoine-Auffret** [comment]
Hi,
I'm trying to pass the ci/circleci: package job but unfortunately it failed.
This job package (1897) failed while installing rpm because it cannot find libarchive13 dependency.
I can reproduce this issue on my environnement using docker and the circleci/ruby:2.3 image.
Nevertheless, the package seems to be available on the debian repositories:
`$ sudo apt list libarchive13
Listing... Done
libarchive13/stable,stable 3.2.2-2+deb9u1 amd64`

`$ sudo apt show libarchive13
Package: libarchive13
Version: 3.2.2-2+deb9u1
Priority: optional
Section: libs
Source: libarchive
Maintainer: Debian Libarchive Maintainers <ah-libarchive@debian.org>
Installed-Size: 793 kB
Depends: libacl1 (>= 2.2.51-8), libbz2-1.0, libc6 (>= 2.16), liblz4-1 (>= 0.0~r130), liblzma5 (>= 5.2.2), liblzo2-2, libnettle6, libxml2 (>= 2.7.4), zlib1g (>= 1:1.1.4)
Suggests: lrzip
Homepage: http://www.libarchive.org/
Tag: role::shared-lib
Download-Size: 294 kB
APT-Sources: http://deb.debian.org/debian stretch/main amd64 Packages
Description: Multi-format archive and compression library (shared library)`

Could you have a look @Dieterbe ?
Thanks in advance

**Comment by Antoine-Auffret** [comment]
I found a workaround.
We need to update the current version of the circleci/ruby:2.3 docker image define [here](https://github.com/grafana/carbon-relay-ng/blob/master/.circleci/config.yml#L36) which hasn't been updated for 2 years.
The minimum version currently working seems to be circleci/ruby:2.5.
I'm updating the ci to change the version of the docker circleci/ruby image to 2.5.

File: route/kafkamdm.go [hunk 1]
```diff
@@ -113,6 +113,15 @@ func NewKafkaMdm(key string, matcher matcher.Matcher, topic, codec, schemasFile,
...
 
 	if saslEnabled {
+		if saslMechanism == "SCRAM-SHA-256" {
+			config.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA256
+			config.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA256} }
+		} else if saslMechanism == "SCRAM-SHA-512" {
+			config.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA512
+			config.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA512} }
+		} else {
+			log.Fatalf("Failed to reconize saslMechanism: '%s'", saslMechanism)
```
Comment (Dieterbe) [inline route/kafkamdm.go:123]: this would break current usage of using sasl without specifying a mechanism.

Comment (Antoine-Auffret) [reply route/kafkamdm.go:123]: Good catch, the following commit fix the issue

File: docs/config.md [hunk 2]
```diff
@@ -235,6 +235,7 @@ tlsSkipVerify  |     N     |  bool       | false   | Whether to skip TLS server
 tlsClientCert  |     N     |  string     | ""      | Client cert for client authentication
 tlsClientKey   |     N     |  string     | ""      | Client key for client authentication
 saslEnabled    |     N     |  bool       | false   | Whether to enable SASL
+saslMechanism  |     N     |  string     | ""      | SASL Mechanism (SCRAM-SHA-256 or SCRAM-SHA-512)
```
Comment (Dieterbe) [inline docs/config.md:238]: let's document what happens if not specified

File: route/kafkamdm.go [hunk 3]
```diff
@@ -113,6 +113,15 @@ func NewKafkaMdm(key string, matcher matcher.Matcher, topic, codec, schemasFile,
...
 
 	if saslEnabled {
+		if saslMechanism == "SCRAM-SHA-256" {
+			config.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA256
+			config.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA256} }
+		} else if saslMechanism == "SCRAM-SHA-512" {
+			config.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA512
+			config.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA512} }
+		} else if saslMechanism != "" {
+			log.Fatalf("Failed to reconize saslMechanism: '%s'", saslMechanism)
```
Comment (Dieterbe) [inline route/kafkamdm.go:123]:
```suggestion
			log.Fatalf("Failed to recognize saslMechanism: '%s'", saslMechanism)
```

File: docs/config.md [hunk 4]
```diff
@@ -235,6 +235,7 @@ tlsSkipVerify  |     N     |  bool       | false   | Whether to skip TLS server
 tlsClientCert  |     N     |  string     | ""      | Client cert for client authentication
 tlsClientKey   |     N     |  string     | ""      | Client key for client authentication
 saslEnabled    |     N     |  bool       | false   | Whether to enable SASL
+saslMechanism  |     N     |  string     | ""      | if empty or not used the SASL Mechanism is disabled, if specified the SASL Mechanism can be SCRAM-SHA-256 or SCRAM-SHA-512
```
Comment (Dieterbe) [inline docs/config.md:238]: > if empty or not used the SASL Mechanism is disabled

this seems misleading. If empty, we default to the "PLAIN" mechanism
How about this:

```suggestion
saslMechanism  |     N     |  string     | ""      | if unset, the PLAIN mechanism is used. You can also specify SCRAM-SHA-256 or SCRAM-SHA-512.
```

we could also allow to explicitly set 'PLAIN' but i don't feel strongly about this.

**Comment by Dieterbe** [comment]
looks good. could you undo the changes to bindata.go so that file remains the same? i realize it's probably our make script that might cause the difference but i'm suspicious of the change with so many deletions and i don't think this file _needs_ to be changed really.

**Comment by Dieterbe** [comment]
nice work @Antoine-Auffret , thanks!

**Comment by Antoine-Auffret** [comment]
@Dieterbe 
There is an issue when circleci is trying to connect to docker when merging branch to master.
https://github.com/grafana/carbon-relay-ng/blob/d3946f9c5bcf4324900da37ae7e2f95b947bb329/.circleci/config.yml#L86
The previous two merge to master failed : [deploy 1931](https://app.circleci.com/pipelines/github/grafana/carbon-relay-ng/146/workflows/848fd8ab-42ef-4866-99c7-d293843136aa/jobs/1931) and [deploy 1961](https://app.circleci.com/pipelines/github/grafana/carbon-relay-ng/151/workflows/63d2a4a5-0fae-4e28-b8fb-c85f4353dbd5/jobs/1961)
`Error response from daemon: Get https://registry-1.docker.io/v2/: unauthorized: incorrect username or password`

Do you have recently (between the October 30, 2020 and today) changed the docker credentials or something like that ?

**Comment by Dieterbe** [comment]
@Antoine-Auffret fixed now. note you have to pull now from grafana/carbon-relay-ng (the raintank repo is now deprecated)

**Comment by Antoine-Auffret** [comment]
> @Antoine-Auffret fixed now. note you have to pull now from grafana/carbon-relay-ng (the raintank repo is now deprecated)

Ok, thank you 👍

**Pull request merged**